lib/Dialect/Datapath/DatapathOps.cpp

```diff
@@ -59,6 +66,323 @@ static void printCompressFormat(OpAsmPrinter &printer, Operation *op,
+  auto carry_delay = sum_delay + 1;
+
+  CompressorBit sum = {sum_val, sum_delay};
+  CompressorBit carry = {carry_val, carry_delay};
+  std::pair<CompressorBit, CompressorBit> fa{sum, carry};
+  ++numFullAdders;
+  return fa;
+}
+
+// Construct a full adder for three 1-bit inputs.
+std::pair<CompressorBit, CompressorBit>
+CompressorTree::halfAdderWithDelay(OpBuilder &builder, CompressorBit a,
+                                   CompressorBit b) {
+  auto sum_val = builder.createOrFold<comb::XorOp>(loc, a.val, b.val, true);
+  auto carry_val = builder.createOrFold<comb::AndOp>(loc, a.val, b.val, true);
+
+  auto sum_delay = std::max(a.delay, b.delay) + 1;
+  auto carry_delay = sum_delay;
+
+  CompressorBit sum = {sum_val, sum_delay};
+  CompressorBit carry = {carry_val, carry_delay};
+  std::pair<CompressorBit, CompressorBit> ha{sum, carry};
+  return ha;
+}
+
+// Map input rows to column representation
+CompressorTree::CompressorTree(const SmallVector<SmallVector<Value>> &addends,
+                               Location loc)
+    : originalAddends(addends), usingTiming(false), numStages(0),
+      numFullAdders(0), loc(loc) {
+  assert(addends.size() > 2);
+  // Number of bits in a row == bitwidth of input addends
+  // Compressors will be formed of uniform bitwidth addends
+  width = addends[0].size();
+  SmallVector<SmallVector<CompressorBit>> initColumns(width);
+  columns = initColumns;
+  // Convert addends rows to columns
+  // Known bits analysis constructs a minimal array - skipping zeros
+  for (auto row : addends) {
+    for (size_t i = 0; i < width; ++i) {
+      CompressorBit bit = {row[i], 0};
+      // TODO: Fold Constant 1s
+      auto knownBits = comb::computeKnownBits(bit.val);
+      if (knownBits.isZero())
+        continue;
+      // Add non-zero bit to the column
+      columns[i].push_back(bit);
+    }
+  }
+}
+
+// Update the input delays based on longest path analysis
+void CompressorTree::withInputDelays(
+    const SmallVector<SmallVector<int64_t>> inputDelays) {
+  assert(inputDelays.size() == originalAddends.size() &&
+         "Input delays must match number of addends");
+  for (size_t i = 0; i < inputDelays.size(); ++i) {
+    assert(inputDelays[i].size() == width &&
+           "Input delays must match bitwidth of addends");
+    for (size_t j = 0; j < width; ++j) {
+      // Find the corresponding bit in the column and update its delay
+      auto it = std::find_if(columns[j].begin(), columns[j].end(),
+                             [val = originalAddends[i][j]](const auto &bit) {
+                               return bit.val == val;
+                             });
+      if (it != columns[j].end())
+        it->delay = inputDelays[i][j];
+    }
+  }
+  usingTiming = true;
+}
+
+size_t CompressorTree::getMaxHeight() const {
+  size_t maxSize = 0;
+  for (const auto &column : columns) {
+    maxSize = std::max(maxSize, column.size());
+  }
+  return maxSize;
+}
```

Comment:
nit: please remove single brace
```suggestion
  for (const auto &column : columns)
    maxSize = std::max(maxSize, column.size());
  return maxSize;
}
```